File: _learning/fcc_algorithms/single_linked.js
class Node {
  constructor(data){
    this.data = data
    this.nextNode = null
  }
}


class LinkedList {
  constructor(){
    this.head = null
    
  }

  isEmpty() {
    return this.head == null
  }

  size(){
    let current = this.head
    let count = 0
    
    while (current != null){
      count++
      current = current.nextNode
    }

    return count
  }

  search(key){
    let current = this.head

    while(current != null){
      if(current.data === key){
        return current
      }

      current = current.nextNode
    }
    return null
  }

  add(data){
    let newNode = new Node(data)
    newNode.nextNode =  this.head
    this.head = newNode
  }

  toString() {
    let nodes = [];
    let current = this.head;
  
    while (current != null) {
      if (current === this.head) {
        nodes.push(`[Head: ${current.data}]`);
      } else if (current.nextNode === null) {
        nodes.push(`[Tail: ${current.data}]`);
      } else {
        nodes.push(`[${current.data}]`);
      }
  
      current = current.nextNode;
    }

  
    console.log(nodes)
    return nodes.join('->').toString();
  }
}


const list = new LinkedList()

console.log(list.isEmpty())
console.log(list.size())
list.add(1)
list.add(2)
list.add(3)
list.add(4)
list.add(5)
console.log(list.search(4))
console.log(list.search(9))

console.log('--')
console.log(list.toString())

File: _learning/fcc_algorithms/single_linked.py
class Node:
    """
    An object for storing a single node of a linked list
    Models two attributes - data and the link to next node in the list
    """
    data = None
    next_node = None

    def __init__(self, data):
        self.data = data

    def __repr__(self):
        return "<Node data: %s>" % self.data

class LinkedList: 
    """
    Single Linked list
    """
    def __init__(self):
        self.head = None


    def __repr__(self):
        """
        Return a string representation of the list
        Takes 0(n) time
        """
        nodes = []
        current = self.head
        
        while current:
            if current is self.head:
                nodes.append (" [Head: %s]" % current.data)
            elif current. next_node is None:
                nodes.append ("[Tail: %s]" % current.data)
            else:
                nodes.append ("[%s]" % current.data)
            
            current = current.next_node

        return '->'.join(nodes)


    def is_empty(self):
        return self.head == None

    def size(self):
        current = self.head
        count = 0
        
        while current:
            count += 1
            current = current.next_node
        
        return count

    def add(self, data):
        new_node = Node(data)
        new_node.next_node = self.head
        self.head = new_node
    
    def insert(self, data, index):
        if index == 0:
            self.add(data)

        current = self.head
        position = index
        new_node = Node(data)

        while position > 1:
            current = current.next_node
            position -= 1

        prev_node = current
        next_node = current.next_node

        prev_node.next_node = new_node
        new_node.next_node = next_node

    def remove(self, key):
        current = self.head
        prev_node = None
        
        while current:
            if current.data == key and current is self.head:
                self.head = current.next_node
                break
            
            elif current.data == key:
                prev_node.next_node = current.next_node
                break

            else:
                prev_node = current
                current = current.next_node
    

    def node_at_index(self, index):
        if index == 0:
            return self.head
        else:
            current = self.head
            position = 0

            while position < index:
                current = current.next_node
                position += 1
        
            return current
        







list = LinkedList()
list.add(1)
list.add(2)
list.add(3)
list.add(5)
list.add(6)
list.insert(4,2)
print(list)
list.remove(6)
print(list)

print(list.is_empty())
print(list.size())


